workshop/6-CustomerDataPlatforms/6.1-Segment.ipynb

```diff
@@ -0,0 +1,303 @@
+    "print(json.dumps(response.json(), indent = 2))"
+   ]
+  },
+  {
+   "source": [
+    "The return value from the recommendation service should look like this:\n",
+    "\n",
+    "```\n",
+    "[\n",
+    "  {\n",
+    "    \"product\": {\n",
+    "      \"id\": \"2\",\n",
+    "      \"url\": \"http://recs.cloudfront.net/#/product/2\",\n",
+    "      \"sk\": \"\",\n",
+    "      \"name\": \"Striped Shirt\",\n",
+    "      \"category\": \"apparel\",\n",
+    "      \"style\": \"shirt\",\n",
+    "      \"description\": \"A classic look for the summer season.\",\n",
+    "      \"price\": 9.99,\n",
+    "      \"image\": \"http://recs.cloudfront.net/images/apparel/1.jpg\",\n",
+    "      \"featured\": \"true\"\n",
+    "    }\n",
+    "  },\n",
+    "  {\n",
+    "    \"product\": {\n",
+    "      \"id\": \"1\",\n",
+    "      \"url\": \"http://recs.cloudfront.net/#/product/1\",\n",
+    "      \"sk\": \"\",\n",
+    "      \"name\": \"Black Leather Backpack\",\n",
+    "      \"category\": \"accessories\",\n",
+    "      \"style\": \"bag\",\n",
+    "      \"description\": \"Our handmade leather backpack will look great at the office or out on the town.\",\n",
+    "      \"price\": 109.99,\n",
+    "      \"image\": \"http://recs.cloudfront.net/images/accessories/1.jpg\",\n",
+    "      \"featured\": \"true\"\n",
+    "    }\n",
+    "  },\n",
+    "\n",
+    "  ...\n",
+    "]\n",
+    "```\n",
+    "\n",
+    "Setting up a similar service in your deployment of Amazon Personalize will give you the ability to surface product recommendations directly into Segment to power other services that require real-time user personalization such as a mobile application or a web application.  \n",
+    "\n",
+    "The recommendation service deployed in the workshop queries a Personalize campaign with the `user_id` that you provide.  Personalize will return a list of recommended product IDs for the user.  The service then looks up the catalog ID for each product and decorates the response with the product URL, image, name, etc.  These attributes should be available in your catalog management system.  In the case of the Retail Demo Store, there is a catalog service that provides (simulated) catalog items for the workshop.  The advantage of a service like this is that your target services like Segment do not need to do an additional lookup in order to get the information they require in order to surface a recommendation to a user, such as product name, product category and the image associated with the product.\n",
+    "\n",
+    "Next, we will use the aforementioned service to pull our demo user recommendations and push them into Segment. We’re going to install Segment’s Python library, then create demo user profiles within Segment and pull in recommendations for each user from Amazon Personalize:\n"
```

Comment:
At some point it would be good to describe how customers can scale updating Personas with recommendations from Personalize. The inference Lambda function is a great approach since it keeps the recommendations super fresh based on each user's latest activity. However, it's transactional (and maybe an anti-pattern for a Personas integration?). Another approach is to use Personalize [Batch Recommendations](https://docs.aws.amazon.com/personalize/latest/dg/recommendations-batch.html) to generate recommendations for all/many users at once for period updates in Personas. Is there an analog batch update process for Personas that can be used to sync batch recommendations from Personalize? Just a thought.

Reply:
This is a really important point.  The transactional approach is an OK pattern for Personas - it will only send updates to users that have recently been added to an audience, and that will work for certain use cases, however the batch approach is likely to be much more economical from a cost and performance standpoint, especially for very large audiences.  The Segment team did not have time to dive into this with me but I am pushing them into supporting a batch audiences approach here since this would allow us to use the Personalize batch inference mode.  For this first version of the workshop though we are not going to be able to deploy this.